### PR TITLE
fix: reconcile metrics proto, move `det trial describe` to the new API. [DET-7617]

### PR DESCRIPTION
**Pull request by ioga**
## Description

- `det trial describe --json --metrics` is used to extract detailed info about trial for post-processing. It's been for large trials due to postgres json marshalling limits. This PR moves it to the new, paginated API.
- New API format is different. This is a breaking change for `det trial describe --json` output.
- Move `MetricsWorkload` and `TrialMetrics` protos to use `checkpointv1.Metrics` object for metrics instead of custom disjointed fields. This is a breaking change for REST APIs, but user-facing Core APIs for `report_{training,validation}_metrics` stay the same.
- Allow users to request per-batch metrics from `GetTrialWorkloads` API using `include_batch_metrics` query flag.

## Test Plan

- Tabulated `det trial describe [--metrics]` format should stay the same before previous and new version.
- `det trial describe --json [--metrics]` format will change, but should contain the same volume of information.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/static/srv/proto_get_trial_workloads_with_batch_metrics.sql [hunk 1]
```diff
@@ -0,0 +1,72 @@
+WITH validations_vt AS (
```
Comment (ioga) [inline master/static/srv/proto_get_trial_workloads_with_batch_metrics.sql:1]: TODO: try to make this conditional in the same query as `proto_get_trial_workloads`

File: master/internal/db/postgres_experiments_intg_test.go [hunk 2]
```diff
@@ -430,16 +430,18 @@ func TestCheckpointMetadata(t *testing.T) {
 				m = &trialv1.TrialMetrics{
 					TrialId:        int32(tr.ID),
 					StepsCompleted: stepsCompleted,
-					Metrics: &structpb.Struct{
-						Fields: map[string]*structpb.Value{
-							defaultSearcherMetric: {
-								Kind: &structpb.Value_NumberValue{
-									NumberValue: metricValue,
+					Metrics: &checkpointv1.Metrics{
```
Comment (stoksc) [inline master/internal/db/postgres_experiments_intg_test.go:433]: minor nit, move `checkpointv1.Metrics` now?

File: master/static/srv/proto_get_trial_workloads.sql [hunk 3]
```diff
@@ -31,7 +36,8 @@ checkpoints_vt AS (
         c.report_time as end_time,
         c.uuid,
         c.steps_completed as total_batches,
-        c.resources
+        c.resources,
+        c.metadata
```
Comment (stoksc) [inline master/static/srv/proto_get_trial_workloads.sql:40]: :facepalm:

**Review by stoksc — APPROVED** [review]
lgtm

File: master/internal/db/postgres_experiments_intg_test.go (hunk 2 above)
Comment (ioga) [reply master/internal/db/postgres_experiments_intg_test.go:433]: sorry, do what?

File: harness/determined/cli/util.py [hunk 4]
```diff
@@ -45,3 +46,43 @@
...
+    **kwargs: Any,
+) -> List[Any]:
+    all_objects: List[Any] = []
+    internal_offset = offset or 0
+    while True:
+        r = method(sess, limit=limit, offset=internal_offset, **kwargs)
+        page_objects = getattr(r, agg_field)
+        all_objects += page_objects
+        internal_offset += len(page_objects)
+        if offset is not None or len(page_objects) < limit:
```
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/cli/util.py:86]: yeah. we've ended up with two limit/offset parameters here:

- older one I renamed "fetchone", it is something used in jobs where it fetches only one page
- this one I renamed "fetchall", which fetches all contents page-by-page.

fetchone/fetchall names are from from python DB API for cursors for fetching one/all rows.

This is a bit of a mess from CLI UX perspective. I want to rectify that in a separate PR: add a new flag for controlling if you want to fetch all or single page.

File: master/internal/db/postgres_experiments_intg_test.go (hunk 2 above)
Comment (stoksc) [reply master/internal/db/postgres_experiments_intg_test.go:433]: i just dont know why `Metrics` are in `checkpointv1`.

**Pull request merged**